### PR TITLE
ci: add github build setup configuration

### PR DESCRIPTION
**Pull request by FreeMasen**
This PR adds the ability to add build steps to the local-artifacts job.

A draft until the following items are completed.

- [x] Update all doc comments
- [x] Handle merging workspace configs
- [x] Add validation on job step (`run` and `uses` are not allowed in the same step)
- [x] Reach test coverage on new configuration object

### CODE REVIEW

File: cargo-dist-schema/src/lib.rs [hunk 1]
```diff
@@ -159,6 +159,46 @@ pub struct GithubCiInfo {
     pub external_repo_commit: Option<String>,
 }
 
+/// A github action workflow step
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "kebab-case")]
+pub struct GithubJobStep {
```
Comment (FreeMasen) [inline cargo-dist-schema/src/lib.rs:165]: This repesents the full yml object allowed in [the gh docs](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context) though I am not sure if supporting all items is desired. For example having support for `run` may not be something worth supporting in favor of having users develop gh actions (which can be hosted in the repo directly).

File: cargo-dist/src/tasks.rs [hunk 2]
```diff
@@ -2444,6 +2446,20 @@ impl<'pkg_graph> DistGraphBuilder<'pkg_graph> {
...
+    fn read_build_setup(
+        base: &Utf8PathBuf,
+        cfg: &Option<String>,
+    ) -> DistResult<Vec<GithubJobStep>> {
+        let Some(path) = cfg else {
+            return Ok(Vec::new());
+        };
+        let path = base.join(path);
+        let mut f = File::options().read(true).open(&path).map_err(|e| {
+            panic!("Failed to read build-setup path at {path}: {e}",);
```
Comment (FreeMasen) [inline cargo-dist/src/tasks.rs:2458]: todo: error properly here

File: cargo-dist/tests/snapshots/axolotlsay_generic_workspace_basic.snap [hunk 3]
```diff
@@ -3704,6 +3704,22 @@ jobs:
...
+        uses: "some-action-user/some-other-action"
+        env:
+          "ENV_VAR": "${{ github.head_ref }}"
+      - name: "Some Run Step simple"
+        run: "echo \"hello world!\""
+      - name: "Some Run Step Multi"
+        run: |
+          HW="hello world"
+          echo $HW
+        shell: "bash"
```
Comment (FreeMasen) [inline cargo-dist/tests/snapshots/axolotlsay_generic_workspace_basic.snap:3722]: I'm not exactly sure why this is happening but I can keep looking.

File: book/src/ci/github.md [hunk 4]
```diff
@@ -150,6 +150,73 @@ While the linkage report can be run locally, the report for Linux artifacts can
 The Windows report is currently unable to provide information about the sources of libraries.
 
 
+### Customizing Build Setup
+
+> since 0.20.0
```
Comment (FreeMasen) [inline book/src/ci/github.md:155]: aspirational?

File: cargo-dist/tests/integration-tests.rs [hunk 5]
```diff
@@ -1721,3 +1721,44 @@ fn axolotlsay_generic_workspace_basic() -> Result<(), miette::Report> {
...
+        ))?;
+
+        // Run generate to make sure stuff is up to date before running other commands
+        let ci_result = ctx.cargo_dist_generate(test_name)?;
+        let ci_snap = ci_result.check_all()?;
+        // Do usual build+plan checks
+        let main_result = ctx.cargo_dist_build_and_plan(test_name)?;
+        let main_snap = main_result.check_all(&ctx, ".cargo/bin/")?;
+        // snapshot all
+        main_snap.join(ci_snap).snap();
```
Comment (FreeMasen) [inline cargo-dist/tests/integration-tests.rs:1761]: Should this only be checking the `ci_snap` since it is targeting a CI only feature?

**Review by Gankra — APPROVED** [review]

**Comment by Gankra** [comment]
I'm gonna land this, but if we want to tweak the basedir we can do a follow up.

Thanks so much!!!

**Pull request merged**